### PR TITLE
[https://nvbugs/6566891][fix] Use FlashInfer FA2 for Gemma4 on SM120 and SM121

### PR DESCRIPTION
**Pull request by lfr-0531**
## Dev Engineer Review

- Gemma4 attention now uses `trtllm-gen` only on SM100f.
- Gemma4 uses FlashInfer FA2 on SM120, SM121, and other non-SM100f architectures.
- The routing applies to both sliding-window and full-attention layers.
- Tests cover backend selection and CUDA Graph behavior for both paths.
- No public API or dependency changes were introduced.
- No configuration files or test-list files changed.
- Validation passed, including `pre-commit`, SM120 H256/H512 checks, and a short NVFP4 generation with FP8 KV cache.

**Verdict: sufficient.**

## QA Engineer Review

Modified test functions:

- `test_all_layers_use_trtllm_gen` was renamed to `test_all_layers_use_trtllm_gen_on_sm100f`.
- `test_non_sm100f_layers_use_fa2` was added.
- CUDA Graph backend-selection coverage was updated for SM100f and non-SM100f architectures.

These tests are not listed in `tests/integration/test_lists/` because no test-list files changed. The unit tests provide direct coverage for the updated backend-selection behavior.

**Verdict: sufficient.**

## Description

Gemma4 currently forces every `FlashInferAttention` layer to use the TRTLLM-Gen inner backend. TRTLLM-Gen does not provide the required attention kernels for SM120/SM121, so this path reaches an unsupported-architecture assertion on DGX Spark (SM121).

This change uses FlashInfer FA2 for Gemma4 on SM120 and SM121 while preserving the existing TRTLLM-Gen path on all other architectures. It covers both the H256 sliding-window layers and H512 full-attention layers used by Gemma4. The change is limited to backend selection and does not introduce public API changes or new dependencies.

Related issue: https://nvbugspro.nvidia.com/bug/6566891

Related PR: https://github.com/NVIDIA/TensorRT-LLM/pull/16214

## Test Coverage

- Added backend-selection regression coverage for:
  - SM100: all Gemma4 layers continue to use TRTLLM-Gen.
  - SM120 and SM121: all sliding-window and full-attention layers use FA2.
- Built and installed TensorRT-LLM on an NVIDIA RTX PRO 6000 Blackwell Workstation Edition (SM120) with `CMAKE_CUDA_ARCHITECTURES=120-real`.
- Verified that the imported Python modules and compiled bindings resolved to the modified worktree.
- Ran `python3 -m pytest -q tests/unittest/_torch/modeling/test_modeling_gemma4.py::TestGemma4ModelDefaults`: **7 passed, 2 subtests passed**.
- Loaded the real `nvidia-Gemma-4-26B-A4B-NVFP4` configuration and verified that H256 sliding-window and H512 full-attention layers both selected FA2 on SM120.
- Completed a short end-to-end generation with the real NVFP4 checkpoint, FP8 KV cache, and `max_seq_len=max_num_tokens=6272`. FlashInfer JIT compiled an SM120f FA2 kernel, and the run completed with `E2E_OK`.
- Ran `pre-commit run --all-files` successfully.
- SM121 backend selection is covered by the regression test; SM121 hardware was not available for this validation.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_gemma4.py [hunk 1]
```diff
@@ -328,14 +329,11 @@ def __init__(
...
-        # backend so that QkvE4m3OBfloat16 context cubins can be used
-        # (context cubins require same Q/KV dtype; decode cubins support
-        # mixed dtypes natively).  Uniform backend avoids workspace
-        # corruption between different wrapper types under CUDA graphs.
-        self.attn.flashinfer_backend = "trtllm-gen"
+        # trtllm-gen does not provide SM120/SM121 kernels. FlashInfer FA2
+        # supports Gemma4 H256/H512, sliding-window attention, and FP8 KV cache
+        # on these architectures. Keep the existing trtllm-gen routing on all
+        # other architectures.
+        self.attn.flashinfer_backend = "fa2" if get_sm_version() in (120, 121) else "trtllm-gen"
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_gemma4.py]: This enumerates the arches known to fail rather than the arches where trtllm-gen works. trtllm-gen FMHA cubins ship for datacenter Blackwell only, so SM90/SM89 fall into the `else` branch and hit the same `TllmGenFmhaRunner ... Unsupported architecture` throw; the next consumer arch will too. Inverting it makes the generic path the default:

```python
from tensorrt_llm._utils import is_sm_100f
...
self.attn.flashinfer_backend = "trtllm-gen" if is_sm_100f() else "fa2"
```

If fa2 is known-bad for the H512 layers on some pre-Blackwell arch, that's worth stating in the comment — but it argues for an explicit unsupported error there, not for routing it into a backend that has no kernels either.

Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_gemma4.py]: This also changes which prefill path handles custom masks, which the comment doesn't mention. `flashinfer.py:2343` gates the Triton custom-mask prefill on `flashinfer_backend == "trtllm-gen"`, and `flashinfer.py:2396` gates the causal fallback for KV-shared layers the same way. Gemma4 passes `local_attention_mask_data` to the sliding layers when image tokens are present (`[modeling_gemma4.py:1207](https://github.com/NVIDIA/TensorRT-LLM/pull/17557/files#diff-f258c701f35c2275b3539629a0bbfb333db8ea4767da9cad3ce7275a4b1bc9d8R1207)`), so on SM120/121 that mask now goes into flashinfer's native `custom_mask` plan instead — a path this model has never used.

Was a multimodal prefill (image input through `Gemma4ForConditionalGeneration`) run on SM120, or only text generation? If only text, please validate the image path before merge, since that is the entry point in the linked bug's traceback.

File: tests/unittest/_torch/modeling/test_modeling_gemma4.py [hunk 2]
```diff
@@ -3388,7 +3417,7 @@ def _run_cuda_graph_real_headdim(self, config_dict, label=""):
         layers = []
         for info in layers_info:
             kwargs = {}
-            # head_dim>256 needs trtllm-gen (fa2 JIT doesn't support it)
+            # On B200, head_dim>256 needs trtllm-gen.
```
Comment (brnguyen2) [inline tests/unittest/_torch/modeling/test_modeling_gemma4.py]: The comment now says "On B200", but the code below still hardcodes `flashinfer_backend="trtllm-gen"` unconditionally, so this test throws `Unsupported architecture` if anyone runs the file on SM12x — exactly the hardware this PR is enabling. It doesn't break CI today (the file is only listed in `l0_b200.yml`), but a comment change alone doesn't make it arch-correct. Either skip the head_dim>256 cases when `not is_sm_100f()`, or select the backend the model does.

File: tensorrt_llm/_torch/models/modeling_gemma4.py (hunk 1 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/models/modeling_gemma4.py]: Thanks, Brian! Agreed. I changed the routing to use the positive TRTLLM-Gen capability check: Blackwell (`is_sm_100f()`) uses `trtllm-gen`, while other architectures use FA2. I also updated the tests to cover both capability outcomes.

File: tests/unittest/_torch/modeling/test_modeling_gemma4.py (hunk 2 above)
Comment (lfr-0531) [reply tests/unittest/_torch/modeling/test_modeling_gemma4.py]: I updated the real-head-dimension CUDA Graph helper to select TRTLLM-Gen only when `is_sm_100f()` is true; on SM120/SM121 it now follows the model and uses FA2 for H512 as well.

File: tensorrt_llm/_torch/models/modeling_gemma4.py (hunk 1 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/models/modeling_gemma4.py]: I validated the multimodal custom-mask path on SM120 with the real `nvidia/Gemma-4-26B-A4B-NVFP4` checkpoint and an actual image input. The runtime trace confirmed that a non-empty boolean custom mask was planned with FlashInfer FA2 at head dimension 256 (`CUSTOM_MASK_PLAN backend=fa2 head_dim=256`), and generation completed successfully.
